int64 node_id
float64 data
float64 upper_bound
int64 k